### PR TITLE
fl334/available drivers

### PR DESCRIPTION
**Pull request by fengyuli2002**
### Get available drivers

- [x] implemented `/api/drivers/available`

When given a date=YYYY-MM-DD, startTime=hh:mm, and endTime=hh:mm query, this endpoint should return a list of drivers that are available during this date and time range, based on a driver's availability.

### Test Plan 

I called this endpoint on Postman and it runs successfully.
It is able to correctly filter out unavailable drivers.

### Notes 

1. This endpoint assumes there will not be drivers during weekends.
2. An empty list is returned when there is no drivers available.
3. Another possible output is `{err : 'startTime must precede endTime'}`

### CODE REVIEW

File: server/router/driver.ts [hunk 1]
```diff
@@ -12,6 +12,52 @@ import { UserType } from '../models/subscription';
...
+  }
+
+  db.getAll(res, Driver, tableName, (doc: DriverType[]) => {
+    const drivers = doc.filter((driver) => {
+      const availStart = (() => {
+        if (reqDay === 2) return driver.availability.Tue?.startTime;
+        if (reqDay === 1) return driver.availability.Mon?.startTime;
+        if (reqDay === 3) return driver.availability.Wed?.startTime;
+        if (reqDay === 4) return driver.availability.Thu?.startTime;
+        if (reqDay === 5) return driver.availability.Fri?.startTime;
```
Comment (mg2239) [inline server/router/driver.ts:36]: i think your original approach of getting the day as a string (Mon, Tue, etc.), then finding `driver.availability[day].startTime` or `endTime` was better. doing it like this results in a lot of redundant code

File: server/router/driver.ts [hunk 2]
```diff
@@ -12,6 +12,52 @@ import { UserType } from '../models/subscription';
...
+      if (availStart === undefined || availEnd === undefined) {
+        return false;
+      }
+      return (
+        availEnd >= (reqEndTime as string) &&
+        availStart <= (reqStartTime as string)
+      );
+    });
+
+    res.send(drivers);
```
Comment (mg2239) [inline server/router/driver.ts:57]: if the response is an array, the body should be `{ data: drivers }`.

**Review by mg2239 — CHANGES_REQUESTED** [review]
nice job getting this done so quickly! had a few comments below, also if you can remove the `'/:id/:startTime/:endTime'` endpoint that would be great too

File: server/router/driver.ts (hunk 2 above)
Comment (fengyuli2002) [reply server/router/driver.ts:57]: Done

File: server/router/driver.ts [hunk 3]
```diff
@@ -12,6 +12,52 @@ import { UserType } from '../models/subscription';
...
+// Get all available drivers at a specific date and time
+router.get('/available', validateUser('User'), (req, res) => {
+  const { date, startTime: reqStartTime, endTime: reqEndTime } = req.query;
+  const reqDay = moment(date as string).day();
+
+  if ((reqStartTime as string) >= (reqEndTime as string)) {
+    res.status(400).send({ err: 'startTime must precede endTime' });
+  }
+
+  db.getAll(res, Driver, tableName, (doc: DriverType[]) => {
```
[the remark this answers is not in the document]
Comment (fengyuli2002) [reply server/router/driver.ts:29]: No. Dynamoose condition is not able to check if a startTime is greater than another time because not all drivers have startTime on that particular day.

File: server/router/driver.ts (hunk 1 above)
Comment (fengyuli2002) [reply server/router/driver.ts:36]: Done

File: server/router/driver.ts [hunk 4]
```diff
@@ -20,11 +64,6 @@ router.get('/:id', validateUser('User'), (req, res) => {
   db.getById(res, Driver, id, tableName);
 });
 
-// Get all drivers
-router.get('/', validateUser('Admin'), (req, res) => {
-  db.getAll(res, Driver, tableName);
-});
-
```
Comment (mg2239) [inline server/router/driver.ts:27]: why was this removed? we still use this endpoint on the frontend

File: server/router/driver.ts [hunk 5]
```diff
@@ -12,6 +12,50 @@ import { UserType } from '../models/subscription';
...
+// Get all available drivers at a specific date and time
+router.get('/available', validateUser('User'), (req, res) => {
+  const { date, startTime: reqStartTime, endTime: reqEndTime } = req.query;
+  const numToDay = {
+    1: 'Mon',
+    2: 'Tue',
+    3: 'Web',
+    4: 'Thu',
+    5: 'Fri',
+  };
```
Comment (mg2239) [inline server/router/driver.ts:29]: nice change! maybe use an array here instead? theres a little too much typecasting going on for `reqDate`. you can put placeholders (like undefined) for array position 0 and 7

File: server/router/driver.ts [hunk 6]
```diff
@@ -12,6 +12,50 @@ import { UserType } from '../models/subscription';
...
 const tableName = 'Drivers';
 
+// Get all drivers
+router.get('/', validateUser('Admin'), (req, res) => {
+  db.getAll(res, Driver, tableName);
+});
+
+// Get all available drivers at a specific date and time
+router.get('/available', validateUser('User'), (req, res) => {
+  const { date, startTime: reqStartTime, endTime: reqEndTime } = req.query;
```
Comment (mg2239) [inline server/router/driver.ts:22]: since you repeatedly use `reqStartTime as string` and `reqEndTime as string` multiple times, you should create separate string variables for `reqStartTime` and `reqEndTime`

File: server/router/driver.ts [hunk 7]
```diff
@@ -12,6 +12,50 @@ import { UserType } from '../models/subscription';
...
+
+  if ((reqStartTime as string) >= (reqEndTime as string)) {
+    res.status(400).send({ err: 'startTime must precede endTime' });
+  }
+
+  db.getAll(res, Driver, tableName, (doc: DriverType[]) => {
+    const drivers = doc.filter((driver) => {
+      const availStart = !reqDate
+        ? reqDate
+        : (driver.availability as any)[reqDate]?.startTime;
```
Comment (mg2239) [inline server/router/driver.ts:40]: we want to minimize as many `as any` casts as possible, so make sure `reqDay` is typed correctly. you can do this by adding a type to `numToDay`.

Comment (mg2239) [inline server/router/driver.ts:40]: we can simplify this by doing `reqDate && ...`
if `reqDate` is undefined, `availStart` will be undefined.

**Review by mg2239 — CHANGES_REQUESTED** [review]
very close!

File: server/router/driver.ts (hunk 4 above)
Comment (fengyuli2002) [reply server/router/driver.ts:27]: Actually it's not removed. Now it's the first endpoint in driver.ts.

File: server/router/driver.ts (hunk 6 above)
Comment (fengyuli2002) [reply server/router/driver.ts:22]: Done

File: server/router/driver.ts (hunk 5 above)
Comment (fengyuli2002) [reply server/router/driver.ts:29]: Done

File: server/router/driver.ts (hunk 7 above)
Comment (fengyuli2002) [reply server/router/driver.ts:40]: That's nice trick!

Comment (fengyuli2002) [reply server/router/driver.ts:40]: Done, but have to import AvailabilityType.

Comment (mg2239) [reply server/router/driver.ts]: thats fine

File: server/router/driver.ts [hunk 8]
```diff
@@ -4,14 +4,58 @@ import { Condition } from 'dynamoose';
...
+      const availStart = reqDate && driver.availability[reqDate]?.startTime;
+      const availEnd = reqDate && driver.availability[reqDate]?.endTime;
+
+      if (availStart === undefined || availEnd === undefined) {
+        return false;
+      }
+
+      return (
+        availEnd >= (reqEndTime as string) &&
+        availStart <= (reqStartTime as string)
```
Comment (mg2239) [inline server/router/driver.ts:51]: one last thing, remove the `as string` casts for reqEndTime and reqStartTime

**Review by mg2239 — DISMISSED** [review]
really nice job! one last change but should be good to merge after that

**Review by mg2239 — APPROVED** [review]
